File: src/Web/JavaScriptJS/proxy.js
class RemoteDocumentStore {
  constructor() {
    this.fetchCount = 0;
  }

  get(id) {
    this.fetchCount += 1;
    return `doc(${id})`;
  }
}

class DocumentStoreProxy {
  constructor() {
    this.backend = null;
    this.cache = new Map();
  }

  get backendCount() {
    return this.backend === null ? 0 : 1;
  }

  get fetchCount() {
    return this.backend === null ? 0 : this.backend.fetchCount;
  }

  get(id) {
    if (this.cache.has(id)) return this.cache.get(id);
    if (this.backend === null) this.backend = new RemoteDocumentStore();
    const value = this.backend.get(id);
    this.cache.set(id, value);
    return value;
  }
}

const store = new DocumentStoreProxy();
const first = store.get(42);
const second = store.get(42);
console.log(`backend=${store.backendCount};fetches=${store.fetchCount};first=${first};second=${second}`);

File: src/Systems/Nim/proxy.nim
import tables

type
  DocumentStore = ref object of RootObj

  RemoteDocumentStore = ref object of DocumentStore
    fetches: int

  DocumentStoreProxy = ref object of DocumentStore
    backend: RemoteDocumentStore
    cache: Table[int, string]

method getDocument(store: DocumentStore; id: int): string {.base.} =
  raise newException(CatchableError, "abstract document store")

method getDocument(store: RemoteDocumentStore; id: int): string =
  inc store.fetches
  "doc(" & $id & ")"

method getDocument(store: DocumentStoreProxy; id: int): string =
  if store.cache.hasKey(id):
    return store.cache[id]

  if store.backend.isNil:
    store.backend = RemoteDocumentStore(fetches: 0)

  let value = store.backend.getDocument(id)
  store.cache[id] = value
  value

let store = DocumentStoreProxy(cache: initTable[int, string]())
let firstValue = store.getDocument(42)
let secondValue = store.getDocument(42)
let backendCount = if store.backend.isNil: 0 else: 1
let fetchCount = if store.backend.isNil: 0 else: store.backend.fetches

echo "backend=", backendCount, ";fetches=", fetchCount,
  ";first=", firstValue, ";second=", secondValue
